diagnostics/collector.go

```diff
@@ -0,0 +1,170 @@
+package diagnostics
+
+import (
+	"archive/zip"
+	"context"
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"io"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/treeverse/lakefs/db"
+)
+
+type Collector struct {
+	db db.Database
+}
+
+const (
+	maxRecordsPerQueryCollect = 1000
+	csvFileExt                = ".csv"
+)
+
+var ErrNoColumnsFound = errors.New("no columns found")
+
+func NewCollector(adb db.Database) *Collector {
+	return &Collector{
+		db: adb,
+	}
+}
+
+func (c *Collector) Collect(ctx context.Context, w io.Writer) (err error) {
+	writer := zip.NewWriter(w)
+	defer func() { err = writer.Close() }()
+
+	var errs []error
+	contentFromTables := []string{
+		"auth_installation_metadata",
+		"schema_migrations",
+		"pg_stat_database",
+	}
+	for _, tbl := range contentFromTables {
+		err = c.writeTableContent(ctx, writer, tbl)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("write table content for %s %w", tbl, err))
+		}
+	}
+
+	countFromTables := []string{
+		"catalog_branches",
+		"catalog_commits",
+		"catalog_repositories",
+		"auth_users",
+	}
+	for _, tbl := range countFromTables {
+		err = c.writeTableCount(ctx, writer, tbl)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("write table count for %s %w", tbl, err))
+		}
+	}
+
+	err = c.writeRawQueryContent(ctx, writer, "table_sizes", `
+SELECT *, pg_size_pretty(total_bytes) AS total
+    , pg_size_pretty(index_bytes) AS INDEX
+    , pg_size_pretty(toast_bytes) AS toast
+    , pg_size_pretty(table_bytes) AS TABLE
+  FROM (SELECT *, total_bytes-index_bytes-COALESCE(toast_bytes,0) AS table_bytes FROM (
+      SELECT c.oid,nspname AS table_schema, relname AS TABLE_NAME
+              , c.reltuples AS row_estimate
+              , pg_total_relation_size(c.oid) AS total_bytes
+              , pg_indexes_size(c.oid) AS index_bytes
+              , pg_total_relation_size(reltoastrelid) AS toast_bytes
+          FROM pg_class c
+          LEFT JOIN pg_namespace n ON n.oid = c.relnamespace
+          WHERE relkind = 'r') a) a`)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("get table sizes %w", err))
+	}
+
+	// write all errors into log
+	if err := c.writeErrors(ctx, writer, errs); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Collector) writeQueryContent(ctx context.Context, writer *zip.Writer, name string, selectBuilder sq.SelectBuilder) error {
+	q := selectBuilder.Limit(maxRecordsPerQueryCollect)
+	query, args, err := q.ToSql()
+	if err != nil {
+		return fmt.Errorf("query build: %w", err)
+	}
+	return c.writeRawQueryContent(ctx, writer, name, query, args...)
+}
+
+func (c *Collector) writeRawQueryContent(ctx context.Context, writer *zip.Writer, name string, query string, args ...interface{}) error {
```

Comment:
```suggestion
func (c *Collector) writeRawQueryContent(ctx context.Context, zipWriter *zip.Writer, name string, query string, args ...interface{}) error {
```

Reply:
since you have csvWriter...

Reply:
It will be fewer changes to the code if tomorrow I'll add support for tgz and just pass an interface ;)